{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "baf1805b",
   "metadata": {},
   "outputs": [],
   "source": [
    "LIST_LOCATION = './data/lists4node.txt'\n",
    "RESULTS_LOCATION = './data/results4node.txt'\n",
    "VERBOSE = False\n",
    "TO_FILE = True\n",
    "OUTPUT_FILE = 'results.out'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9860fd89",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-03-14 09:53:40.001489: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2023-03-14 09:53:40.503181: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2023-03-14 09:53:40.503236: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2023-03-14 09:53:41.976564: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory\n",
      "2023-03-14 09:53:41.976715: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory\n",
      "2023-03-14 09:53:41.976731: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "np.set_printoptions(precision=3, suppress=True)\n",
    "import os\n",
    "import tensorflow as tf\n",
    "import pandas as pd\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from tensorflow.keras import layers\n",
    "from tensorflow import keras"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8acc088d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Debugging\n",
    "def log(s):\n",
    "    if VERBOSE:\n",
    "        print(s)\n",
    "    if TO_FILE:\n",
    "        if os.path.exists(OUTPUT_FILE):\n",
    "            append_write = 'a'\n",
    "        else:\n",
    "            append_write = 'w'\n",
    "        fh = open(OUTPUT_FILE, append_write)\n",
    "        fh.write(s + '\\n')\n",
    "        fh.close()\n",
    "\n",
    "            \n",
    "\n",
    "\n",
    "# Loading data from the files\n",
    "def preprocess():\n",
    "    data = []\n",
    "    results = []\n",
    "\n",
    "    with open(LIST_LOCATION) as listFile:\n",
    "        with open(RESULTS_LOCATION) as resFile:\n",
    "            for lis, res in zip(listFile, resFile):\n",
    "                intList = list(map(int, lis[:-1].split(',')))\n",
    "\n",
    "                results.append(int(res[:-1]))\n",
    "                data.append(intList)\n",
    "\n",
    "    return pd.DataFrame(data), pd.DataFrame(results)\n",
    "\n",
    "\n",
    "# Tiny model\n",
    "def build_and_compile_model(norm):\n",
    "    model = keras.Sequential([\n",
    "        norm,\n",
    "        layers.Dense(1000, input_dim=4,  activation='relu'),\n",
    "        layers.Dense(500, activation='relu'),\n",
    "        layers.Dense(1, activation='linear')\n",
    "    ])\n",
    "\n",
    "    model.compile(loss='mean_squared_error',\n",
    "                  optimizer='adam')\n",
    "    return model\n",
    "\n",
    "def plot_loss(history):\n",
    "  plt.plot(history.history['loss'], label='loss')\n",
    "  plt.plot(history.history['val_loss'], label='val_loss')\n",
    "  plt.ylim([0, 10])\n",
    "  plt.xlabel('Epoch')\n",
    "  plt.ylabel('Error [MPG]')\n",
    "  plt.legend()\n",
    "  plt.grid(True)\n",
    "\n",
    "def plot_predictions(actual_values, predictions):\n",
    "    a = plt.axes(aspect='equal')\n",
    "    plt.scatter(res_tst.T, test_predictions)\n",
    "    plt.xlabel('True Values [MPG]')\n",
    "    plt.ylabel('Predictions [MPG]')\n",
    "    lims = [0, 50]\n",
    "    plt.xlim(lims)\n",
    "    plt.ylim(lims)\n",
    "    _ = plt.plot(lims, lims)\n",
    "    \n",
    "def plot_error(error):\n",
    "    plt.hist(error.T, bins=25)\n",
    "    plt.xlabel('Percentage error')\n",
    "    plt.ylabel('Count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c1a4f31e",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: './data/lists4nodes.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m data, res \u001b[38;5;241m=\u001b[39m \u001b[43mpreprocess\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m dat_trn, dat_tst, res_trn, res_tst \u001b[38;5;241m=\u001b[39m train_test_split(data, res, test_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.2\u001b[39m, random_state\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m42\u001b[39m)\n\u001b[1;32m      4\u001b[0m normalizer\u001b[38;5;241m.\u001b[39madapt(dat_trn)\n",
      "Cell \u001b[0;32mIn[5], line 8\u001b[0m, in \u001b[0;36mpreprocess\u001b[0;34m()\u001b[0m\n\u001b[1;32m      5\u001b[0m data \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m      6\u001b[0m results \u001b[38;5;241m=\u001b[39m []\n\u001b[0;32m----> 8\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mLIST_LOCATION\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m listFile:\n\u001b[1;32m      9\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(RESULTS_LOCATION) \u001b[38;5;28;01mas\u001b[39;00m resFile:\n\u001b[1;32m     10\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m lis, res \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(listFile,resFile):\n",
      "File \u001b[0;32m/usr/lib/python3.10/site-packages/IPython/core/interactiveshell.py:282\u001b[0m, in \u001b[0;36m_modified_open\u001b[0;34m(file, *args, **kwargs)\u001b[0m\n\u001b[1;32m    275\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m file \u001b[38;5;129;01min\u001b[39;00m {\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m1\u001b[39m, \u001b[38;5;241m2\u001b[39m}:\n\u001b[1;32m    276\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    277\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIPython won\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mt let you open fd=\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfile\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m by default \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    278\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mas it is likely to crash IPython. If you know what you are doing, \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    279\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124myou can use builtins\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m open.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    280\u001b[0m     )\n\u001b[0;32m--> 282\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mio_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: './data/lists4nodes.txt'"
     ]
    }
   ],
   "source": [
    "# Load the data\n",
    "data, res = preprocess()\n",
    "normalizer = tf.keras.layers.Normalization(axis=-1)\n",
    "\n",
    "eval_results = []\n",
    "abs_errors = []\n",
    "\n",
    "data_train, full_test, result_train, full_res = train_test_split(data, res, test_size=0.15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50bb48e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Take data in 20% increments\n",
    "for i in range(15, 100, 20):\n",
    "    # Get the current segment\n",
    "    test_sz = i/100\n",
    "    log(\"Current test size is: \" + str(test_sz))\n",
    "    _, data_slice, _, result_slice = train_test_split(data_train, result_train, test_size=test_sz)\n",
    "\n",
    "    # Train/Test split\n",
    "    log(\"Data size - training = \" + str(len(data_slice)))\n",
    "\n",
    "    # Build model\n",
    "    normalizer.adapt(data_slice)\n",
    "    dnn_model = build_and_compile_model(normalizer)\n",
    "\n",
    "    # Train\n",
    "    history = dnn_model.fit(\n",
    "        data_slice,\n",
    "        result_slice,\n",
    "        validation_split=0.2,\n",
    "        verbose=0, epochs=100)\n",
    "\n",
    "\n",
    "    # Eval\n",
    "    eval_res = dnn_model.evaluate(full_test, full_res, verbose=0)\n",
    "    log(\"Eval results for test size \" + str(test_sz) + \" = \" + str(eval_res))\n",
    "    eval_results.append(eval_res)\n",
    "\n",
    "    # Get error\n",
    "    test_predictions = dnn_model.predict(full_test).flatten()\n",
    "    error = (100 * (test_predictions - full_res.T)) / full_res.T\n",
    "    abs_error = error.T.sum() / len(full_test)\n",
    "    log(\"Error for test size \" + str(test_sz) + \" = \" + str(float(abs_error)))\n",
    "\n",
    "    abs_errors.append(abs_error)\n",
    "    \n",
    "    # Plotting\n",
    "    # plot_loss(history)\n",
    "    # plot_predictions(test_predictions, full_res)\n",
    "    # plot_error(error)\n",
    "\n",
    "print(\"Absolute errors:\\n\")\n",
    "print(abs_errors)\n",
    "print(\"\\n\\n\")\n",
    "\n",
    "print(\"Evaluations:\\n\")\n",
    "print(eval_results)\n",
    "print(\"\\n\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
